### ISSUE TITLE
md-autocomplete

### ISSUE THREAD

**Issue by jelbourn** [angular/material2#117]
Blocked on both #113 and #84

**Comment by pramttl** [comment]
Any updates on this?

**Comment by DevVersion** [comment]
This shouldn't be blocked anymore I guess.

**Comment by michaeljota** [comment]
Hi! If both issues are close, how this is not being updated? Is it being developed?

**Comment by EladBezalel** [comment]
It's in early stages of development

**Comment by michaeljota** [comment]
Thanks you! 😄

**Comment by AndrewWeb-Dev** [comment]
Is regarding autocomplete support on popular text editors?

**Comment by topherfangio** [comment]
This issue is in regards to the Material Autocomplete which you can view at https://material.google.com/components/text-fields.html#text-fields-auto-complete-text-field.

**Comment by gozek** [comment]
is there an estimation regarding the release ? thanks a lot!

**Comment by nkwood** [comment]
@EladBezalel @topherfangio Is there anything on md-autocomplete that might be good for community contribution at this time? I didn't see anything on md-autocomplete in the [design documents](https://github.com/angular/material2/wiki/Design-doc-directory), except for the references in [the design doc for md-input](https://docs.google.com/document/d/1yBbdxDv_Ycer3E_dvstX0h9FJhiddBHX20HODgrUBl8/preview) and the hooks in md-input.

I think there may be a few people that are willing to contribute here, but if there has already been significant design progress that hasn't made it into the publicly available design docs it doesn't make any sense to duplicate that effort.

**Comment by topherfangio** [comment]
@nkwood We are currently still in the design phase for this component and are trying to come to a consensus on the structure and API we plan/want to support.

**Comment by creativedeveloper-net** [comment]
if someone needs autocomplete made in Angular 2 you can have a look here:
https://github.com/creativedeveloper-net/angular2-autocomplete
here is a plunker demo:
https://plnkr.co:443/2aL2Ju
I've made it based on examples from here:
http://stackoverflow.com/questions/35881815/implementing-autocomplete-for-angular2/

Feel free to update/improve it

**Comment by michaeljota** [comment]
@creativedeveloper-net I can see the list of the possible values, however I'm unable to select any.

**Comment by creativedeveloper-net** [comment]
@michaeljota thanks for the input, could you please add it as an issue here:
https://github.com/creativedeveloper-net/angular2-autocomplete/issues
and please describe which browser do you use etc thanks in advance

**Comment by alan345** [comment]
is there an estimation regarding the release ? thanks a lot!

**Comment by kkjf02** [comment]
is there an estimation regarding the release ? thanks a lot! (2)

**Comment by daBishMan** [comment]
is there going to have the ability to add new items that are not part of the suggestions ?  or you can only select what is in the "Suggestions".  I would prefer if we have the ability to do both?

**Comment by MarcusCalidus** [comment]
it would be cool if one could bind to the current input value. Currently I am using md2-autocomplete by @dharmeshpipariya 
But since it is not possible to get the input value from there I use jquery events 
`userChange(event: JQueryEventObject) {
        if (event) {
            this.userInputText = $(event.currentTarget).val().toLowerCase();
            this.onChange(event);
        }
    }`

To have a "new item" item would be awesome like here: https://material.angularjs.org/latest/demo/autocomplete

**Comment by kara** [comment]
https://docs.google.com/document/d/1c9e9-5HlOFKqmdq147yugAQroMLfOeGdRGyB1JGxKVE/preview

**Comment by critchie** [comment]
Please consider the ability to group the results that appear in the drop down.  I have several business cases where one search input is used to search several domain entities and the users want the results grouped by their domain type.

**Comment by bendg25** [comment]
Is there an early release of this available?

**Comment by istiti** [comment]
any ETA?

**Comment by istiti** [comment]
I'm trying to implement autocomplete (for personal use):
https://gist.github.com/istiti/225c29afee646da66224b3388c9b52d9#file-autocomplete-directive-ts-L53

Why I can't change my input model from my directive ? How can I achieve this ? 

Thanks

**Comment by ehegnes** [comment]
@bendg25 Please consider tracking changes to the code to determine release progress. [This GitHub search query may help you.](https://github.com/search?utf8=%E2%9C%93&q=repo%3Aangular%2Fmaterial2+autocomplete&type=Commits&ref=searchresults)

@istiti, @alan345, @kkjf02  ETA inquiries are generally not conducive to improving development efforts. 

@istiti With respect to your autocomplete implementation, please consider posting such questions to another platform (Stack Overflow, for example). Your implementation is not germane to this issue.

**Comment by kara** [comment]
Also worth noting anyone who's interested can want to follow the autocomplete board [here](https://github.com/angular/material2/projects/7).

**Comment by alan345** [comment]
@kara Thanks Kara

**Comment by istvanszoboszlai** [comment]
Hello, the material2 repo seams quite silent. Is there an ETA for this?

**Comment by istiti** [comment]
It's always silent even with new employees going team !!!

**Comment by jelbourn** [comment]
I wouldn't exactly call [the activity on master](https://github.com/angular/material2/commits/master) or [on the autocomplete branch](https://github.com/angular/material2/commits/autocomplete) "silent".

**Comment by alirezamirian** [comment]
I really love how in Material 2 Autocomplete is not a black box component hiding the underlying input.
I was about to add a link to [this issue](https://github.com/angular/material/issues/5040), and then I saw [this commit](https://github.com/angular/material2/commit/3d236a). 
Thank you guys 👍

**Comment by istvanszoboszlai** [comment]
Guys, you can add thumbs down to the questions about ETA.

I really appreciate the hard work you guys put in Material 2, and I did not intend to hurt your feelings, I just really would like to know your plans on releasing autocomplete.

I guess Google is not doing Angular only out of generosity. Google wants developers to use their stack instead of React, and maybe even contribute code to Angular.
Angular 2 without Material 2 is simply not production ready. Material is THE ui recommendation of Google. This is why I (and many another developer) thinks that Material 2 library would need more resources.

@EladBezalel, honestly how could I understand your project management and generally just how the Material 2 team works, and what you work on? I based my view about autocomplete being silent on the kanban board you maintain for this feature.

We have projects that clients are waiting for. One happens to be blocked on autocomplete. I have to tell them something. So guys, do you have an estimated ETA for autocomplete? :)

**Comment by michaeljota** [comment]
@istvanszoboszlai Why? Why even ask for an ETA? Every of us have their own schedule (Even Angular team, and it is public, you can track their doing in the "Milestone" section of the issues), but you can't count with something that does not exist. 

Beside the one mention to the "Milestone" section, the readme explicitly show this as a BETA, and it should NOT be used in production. I use it. You use it. Most of us are using it. But, we have to be grown up here, this is not yet ready. We have to wait. There are plenty of choice if you need an UI framework. 

I'm sorry if I'm being rude, but when I get a notification about this I'm expecting that update to say "It's available guys, test it", but instead a found another post about "They should say when it will be ready", spoiler: They don't know. They are people, like us.

You can help. The code it's available, glob, there is even a branch just for the autocomplete. 😄

**Comment by istvanszoboszlai** [comment]
@michaeljota, I think you were definitely not rude.
Every software team has plans. I am simply curious about that. Maybe I miss something, but I cannot find up-to-date plans on this project. The milestones section is not up-to-date. So I cannot rely on that.
I am using Material2 with https://github.com/Teradata/covalent right now. They maintain nice documentation on their progress, and if I ask ETA than they simply answer.

@michaeljota, what I am really interested in: Which Angular2 UI framework do you think is there to choose? I know of none that native Angular 2.

**Comment by michaeljota** [comment]
Like I say, they have a branch about autocomplete: 

> `@jelbourn`: I wouldn't exactly call the activity on master or on the [autocomplete branch](https://github.com/angular/material2/commits/autocomplete) "silent".

For the other UI for Angular, one of the most complete support would be NgPrime, but for a most complete list, in Angular's website, in the [resources](https://angular.io/resources/) section you can find many libraries for UI composition.

**Comment by istvanszoboszlai** [comment]
Sorry for offtopic posts, this is my last.
I did do my homework about half a year ago and there is not much change in the UI framework landscape.
PrimeNG is an Angular 2 wrapper over their jquery based UI elements. Not exactly native Angular 2.
On the other hand, look, even react has a mature, and awesome material UI library: http://www.material-ui.com/

I think Material 2 is the most important entry point for new users of Angular 2, and still - as awesome as it is - it is not treated as an extremely important project inside Google. This is pretty much a pity, IMHO.

**Comment by istiti** [comment]
@istvanszoboszlai reactmeaterial is 1.5year older than material2. but i agree with you  !mateirla2 team need more resources-employes or so!!

**Comment by michaeljota** [comment]
You should notice that material-ui it's just an open source project, without any big company behind. So, why do you request more resources in an open source project with a big company behind? Instead you should just support the project, if you like. I think it's kind of unfair to request "more devs" from "google" cause this is a "google project". We all are just devs here. This still just another OSS. So, act alike.

**Comment by istiti** [comment]
Lol open source=open ok but when I start a project using this techno or when I join a team using this techo we aren't so free as you thnk! We depends! Before mask all behind opensource explain why these project are open why android is open if no profits?! stop telling opensource here! Its troll. I continue to think here need more resources to fight and be the most used fw

**Comment by ehegnes** [comment]
@jelbourn I hope I am not out of line when requesting this, but may this issue kindly be [locked to collaborators only](https://help.github.com/articles/locking-conversations/)? This way the issue may constructively track progress without the overhead of accumulating inflammatory remarks.

**Comment by kara** [comment]
Update:  The initial version of autocomplete is getting pretty close. At this point, it's bare bones but usable. We are just ironing out any last bugs that come up. I can't give you an exact date when it will be available, but it will definitely go out with our next beta release (soon!). 

@ehegnes I try to avoid closing threads because we do like hearing from the community, but I think you're right that this one is getting derailed a bit.  I'm going to lock it for now. But if anyone wants to comment on potential bugs or feature requests for the autocomplete, please feel free to file a separate issue.

**Comment by EladBezalel** [comment]
@kara i know it's locked but i have to reply to this one -

@istvanszoboszlai @istiti you should listen to @michaeljota,
I joined the material team a year and a half ago after contributing for more than a year to AngularJS material, I did it on my own free time and if I had an issue or a feature suggestion i opened issues and did PRs so not only i'd benefit from it, the entire community will.

It's **super** rude and egoistic to say - *Google should put more devs on that cause they're probably using it* - this project, as the Angular project is open-source, if you use it, contribute! don't demand that some company would throw more money over your problems! 

If the company you work for doesn't understand the open-source community and how it's about to share and contribute, you should teach them or leave.

I get why people ask for ETA, but i don't get why some take this project or the people working for it as granted.

**Comment by kara** [comment]
Closing this, as the initial version of autocomplete is out in beta.2. We'll track future features in separate issues.
